### PR TITLE
KAFKA-4296: Fix LogCleaner statistics rolling

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
ping @ijuma (when you have time)

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 1]
```diff
@@ -378,11 +375,13 @@ private[log] class Cleaner(val id: Int,
...
    * @param map The offset map to use for cleaning segments
    * @param deleteHorizonMs The time to retain delete tombstones
+   * @param stats Collector for cleaning statistics
    */
   private[log] def cleanSegments(log: Log,
                                  segments: Seq[LogSegment], 
                                  map: OffsetMap, 
-                                 deleteHorizonMs: Long) {
+                                 deleteHorizonMs: Long,
+                                 stats: CleanerStats = new CleanerStats()) {
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: Do we want the default? Generally, it's better to avoid defaults like this as it can lead to bugs where one forgets to pass the `stats` instance. Same for other similar cases.

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 2]
```diff
@@ -519,44 +523,43 @@ private[log] class Cleaner(val id: Int,
   private def compressMessages(buffer: ByteBuffer,
                                compressionCodec: CompressionCodec,
                                messageFormatVersion: Byte,
-                               messageAndOffsets: Seq[MessageAndOffset]) {
+                               messageAndOffsets: Seq[MessageAndOffset]): Int = {
     require(compressionCodec != NoCompressionCodec, s"compressionCodec must not be $NoCompressionCodec")
-    if (messageAndOffsets.nonEmpty) {
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: Why did you remove this `if`? The methods below assume that the `messageAndOffsets` are not empty. If you are sure that we never call it with an empty Seq, then we should probably add a `require` to make that explicit.

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala [hunk 3]
```diff
@@ -137,10 +137,38 @@ class CleanerTest extends JUnitSuite {
                (0 until leo.toInt by 2).forall(!keys.contains(_)))
   }
 
+  def testLogCleanerStats(): Unit = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:146]: I assume this test fails without the changes?

**Review by ijuma — COMMENTED** [review]
A few minor comments, looks good otherwise. A bit odd that the existing code went through such great lengths to avoid allocating a cheap object instance per cleaner run.

File: core/src/main/scala/kafka/log/LogCleaner.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: Ack. I thought it seemed odd in an exposed method, so this was a weak attempt to conceal the oddness. I guess it's only exposed for testing though, so maybe it's not too bad.

File: core/src/main/scala/kafka/log/LogCleaner.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: Sure, I'll add the `require` (the only usage of this method is in this class). It was actually a bug before because we would increment the `messagesWritten` counter even if there were no retained messages.

Comment (ijuma) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: Are you sure there was a bug? I'm not sure I see it. `compressMessages` was a no-op if `messageAndOffsets` was empty right?

Comment (hachikuji) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: Actually you're right. I was probably thinking of a bug I introduced when I pulled the `CleanerStats` updating out of this function 😄

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:146]: Not really, this was mainly for completeness (we didn't seem to have much testing around cleaner stats). I guess we can add another test which asserts that `lastStats` is not affected by a call to `clean()`, but it seems like overkill given the refactor. What do you think?

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 4]
```diff
@@ -550,7 +556,7 @@ private[log] class Cleaner(val id: Int,
           for (messageAndOffset <- messageAndOffsets) {
             offset = messageAndOffset.offset
             val innerOffset = if (magic > Message.MagicValue_V0)
-              // The offset of the messages are absolute offset, compute the inner offset.
+            // The offset of the messages are absolute offset, compute the inner offset.
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: Is this intentional?

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala (hunk 3 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:146]: I think it would be nice to have, if not too awkward. If it is too awkward, then OK to skip it since this fixes a bug and improves coverage compared to what we had before.

**Review by ijuma — APPROVED** [review]
Thanks for the PR, LGTM. Left a couple of comments, one of them trivial and the other is more of a question.

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:146]: @ijuma It is a bit awkward. We need to either work through `LogCleaner` which would make the test asynchronous, or we need to expose `CleanerThread` which is nested in `LogCleaner`. It would be nice to get some more coverage in here, but I'm sort of inclined to leave this for future work.

Comment (ijuma) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:146]: Sounds fair.
